Mathlib/Geometry/Manifold/Bordisms.lean

```diff
@@ -0,0 +1,211 @@
+
+instance {n : ℕ} {s : SingularNManifold X n} : NormedAddCommGroup s.E := s.normedAddCommGroup
+
+instance {n : ℕ} {s : SingularNManifold X n} : NormedSpace ℝ s.E := s.normedSpace
+
+instance {n : ℕ} {s : SingularNManifold X n} : TopologicalSpace s.M := s.topSpaceM
+
+instance {n : ℕ} {s : SingularNManifold X n} : TopologicalSpace s.H := s.topSpaceH
+
+instance {n : ℕ} {s : SingularNManifold X n} : ChartedSpace s.H s.M := s.chartedSpace
+
+instance {n : ℕ} {s : SingularNManifold X n} : SmoothManifoldWithCorners s.model s.M := s.smoothMfd
+
+instance {n : ℕ} {s : SingularNManifold X n} : CompactSpace s.M := s.compactSpace
+
+instance {n : ℕ} {s : SingularNManifold X n} : BoundarylessManifold s.model s.M := s.boundaryless
+
+instance {n : ℕ} {s : SingularNManifold X n} : FiniteDimensional ℝ s.E := s.findim
+
+namespace SingularNManifold
+
+variable {n : ℕ}
+
+/-- A map of topological spaces induces a corresponding map of singular n-manifolds. -/
+-- This is part of proving functoriality of the bordism groups.
+noncomputable def map (s : SingularNManifold X n)
+    {φ : X → Y} (hφ : Continuous φ) : SingularNManifold Y n where
+  model := s.model
+  f := φ ∘ s.f
+  hf := hφ.comp s.hf
+  dimension := s.dimension
+
+@[simp]
+lemma map_f (s : SingularNManifold X n) {φ : X → Y} (hφ : Continuous φ) :
+    (s.map hφ).f = φ ∘ s.f :=
+  rfl
+
+lemma map_comp (s : SingularNManifold X n)
+    {φ : X → Y} {ψ : Y → Z} (hφ : Continuous φ) (hψ : Continuous ψ) :
+    ((s.map hφ).map hψ).f = (ψ ∘ φ) ∘ s.f := by
+  simp [Function.comp_def]
+  rfl
+
+-- Let M, M' and W be smooth manifolds.
+universe u
+variable {E E' E'' E''' H H' H'' H''' : Type u} [NormedAddCommGroup E] [NormedSpace ℝ E]
+  [NormedAddCommGroup E'] [NormedSpace ℝ E'] [NormedAddCommGroup E'']  [NormedSpace ℝ E'']
+  [NormedAddCommGroup E'''] [NormedSpace ℝ E''']
+  [TopologicalSpace H] [TopologicalSpace H'] [TopologicalSpace H''] [TopologicalSpace H''']
+
+variable {M : Type u} [TopologicalSpace M] [ChartedSpace H M]
+  {I : ModelWithCorners ℝ E H} [SmoothManifoldWithCorners I M]
+  {M' : Type u} [TopologicalSpace M'] [ChartedSpace H' M']
+  {I' : ModelWithCorners ℝ E' H'} [SmoothManifoldWithCorners I' M'] {n : ℕ}
+  [BoundarylessManifold I M] [CompactSpace M] [FiniteDimensional ℝ E]
+  [BoundarylessManifold I' M'] [CompactSpace M'] [FiniteDimensional ℝ E']
+
+variable (M) in
+/-- If `M` is `n`-dimensional and closed, it is a singular `n`-manifold over itself. -/
+noncomputable def refl (hdim : finrank ℝ E = n) : SingularNManifold M n where
+  H := H
+  model := I
+  dimension := hdim
+  f := id
+  hf := continuous_id
+
+/-- If `(N, f)` is a singular `n`-manifold on `X` and `M` another `n`-dimensional smooth manifold,
+a smooth map `φ : M → N` induces a singular `n`-manifold structure `(M, f ∘ φ)` on `X`. -/
+noncomputable def comap [h: Fact (finrank ℝ E = n)]
+    (s : SingularNManifold X n)
+    {φ : M → s.M} (hφ : Smooth I s.model φ) : SingularNManifold X n where
```

Comment:
It's a little bit strange to assume that phi is smooth, while continuous would do here, no?